### PR TITLE
[Makefile] Refactor CUDA makefile and add Hopper (SM90) to default build

### PR DESCRIPTION
**Pull request by yaox12**
## Description

- Add Hopper (SM90) to default build for CUDA >= 11.8
- Drop support for compute capability < 3.5 (follow PyTorch's convention)
- Correctly set `__cuda_arch_ptx` to be the latest arch in `__cuda_arch_bin`
- Add `Turing`/`Ada`/`Hopper` to the selectable arch list.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by chang-l** [comment]
I am not very familiar with this PR...
I'll try my best to review/comment here, over the weekend or early next week. But I will leave it to @TristonC for approval.

File: cmake/modules/CUDA.cmake [hunk 1]
```diff
@@ -63,10 +70,14 @@ set(CUDA_gpu_detect_output "")
...
       # The line below prevents CMake from inserting a variable with line
       # breaks in the cache
-      message(STATUS "Found CUDA arch ${__nvcc_out}")
+      message(STATUS "Found GPU arch ${__nvcc_out}")
       string(REGEX MATCH "([1-9].[0-9])" __nvcc_out "${__nvcc_out}")
-      string(REPLACE "2.1" "2.1(2.0)" __nvcc_out "${__nvcc_out}")
-      set(CUDA_gpu_detect_output ${__nvcc_out} CACHE INTERNAL "Returned GPU architetures from mshadow_detect_gpus tool" FORCE)
+      if(__nvcc_out VERSION_LESS "3.5")
+        # drop support for cc < 3.5 and build for all known archs.
+        message(WARNING "GPU arch is too old to support.")
```
Comment (TristonC) [inline cmake/modules/CUDA.cmake:77]: Recommend changing the error message to "GPU arch less than 3.5 is not supported."

File: cmake/modules/CUDA.cmake [hunk 2]
```diff
@@ -10,9 +10,16 @@ endif()
 include(CheckCXXCompilerFlag)
 check_cxx_compiler_flag("-std=c++14"   SUPPORT_CXX14)
 
-set(dgl_known_gpu_archs "35 50 60 70")
+set(dgl_known_gpu_archs "35" "50" "60" "70")
```
Comment (TristonC) [inline cmake/modules/CUDA.cmake:13]: What is the reason to change "35 50 60 70" to "35 50 60 70"?

Comment (yaox12) [reply cmake/modules/CUDA.cmake:13]: To make it a list and then in 
https://github.com/dmlc/dgl/blob/3e8af7d66d5dce631f1bd6c5e1176022c15b5f7a/cmake/modules/CUDA.cmake#L157-L159
we can easily get the last item in `__cuda_arch_bin` as `__cuda_arch_ptx`.

File: cmake/modules/CUDA.cmake (hunk 1 above)
Comment (yaox12) [reply cmake/modules/CUDA.cmake]: Changed accordingly.

**Comment by TristonC** [comment]
LGTM

**Review by TristonC — APPROVED** [review]

**Pull request merged**